### DISCUSSION TITLE
How connect IOS app to Internet?

### DISCUSSION THREAD

**Discussion by ZhorovEgor** [airsdk/Adobe-Runtime-Support#3243] [General]
Good day! The situation is as follows: when testing the application through TestFlight, my application cannot connect to the network. Considering that this application has already been properly created for Android using Adobe Animate AIR SDK by Harman, can you tell me what and where to add it?

![Screenshot_1](https://github.com/airsdk/Adobe-Runtime-Support/assets/63095086/5385ea44-cdd3-4927-a6f7-96b19c4961d8)
![Screenshot_2](https://github.com/airsdk/Adobe-Runtime-Support/assets/63095086/fffb0b15-54ab-4326-b6b1-6cf9b9b561d8)
![Screenshot_3](https://github.com/airsdk/Adobe-Runtime-Support/assets/63095086/1a4024ef-bc9a-4de9-a8d0-b4387dc3df37)

**Comment by ajwfrost** [comment]
AIR SDK used to automatically inject the 'internet' permission .. but that was removed because of some other requirements (see #2469)

So please add the permissions element in to your application descriptor for this... (see also #2525 - we do it for you automatically just on debug builds..)

thanks

**Reply by ZhorovEgor** [reply]
You gave comments about the android component, which I'm fine with. I have a problem with ios specifically.

**Reply by ajwfrost** [reply]
Oh, sorry! Too hasty in reading and making assumptions...

If you're trying to connect via HTTP, then I would think it's likely to be related to the iOS app transport security .. see the final comment/suggestion on the below thread:
https://community.adobe.com/t5/air-discussions/why-is-urlloader-in-adobe-air-ios-app-no-more-working/td-p/10145145

thank

**Reply by ZhorovEgor** [reply]
Hi! My application is essentially not knocking on any http resource. It works on the principle of a controller, the application performs the function of a controller for smart stairs (lighting, etc.)

**Reply by ajwfrost** [reply]
Hi

Sorry, I'm not sure what you're trying to do here / how you're doing it. You mention it's a problem connecting to the internet, but not using HTTP? A smart home controller would presumably either be trigging things via an internet-based server (MQTT?) or directly on the local network...?

What ActionScript classes/calls are you making which are failing? Any error messages or failure codes/errors happening on the ActionScript side?

thanks

**Reply by ZhorovEgor** [reply]
The Internet is required in order for the application to save the user's settings in the database. It's not critical yet. The most important thing for us is to give him permission to work in local networks. As you can see in the screenshot, we have displayed logging in the form of text to display the connection status. It shows a connection error.

**Reply by ajwfrost** [reply]
My guess would still be there's some iOS security feature that's blocking it ..
Is there any way you can create a simple test app that reproduces the problem that we could then use here to investigate? 

Looking at a few ideas based on your screenshots:
- can you add an InfoAdditions key of [`NSLocalNetworkUsageDescription`?](https://developer.apple.com/documentation/bundleresources/information_property_list/nslocalnetworkusagedescription?language=objc)
- in case you're using an ANE and implementing things with Bonjour services, see also [`NSBonjourServices`](https://developer.apple.com/documentation/bundleresources/information_property_list/nsbonjourservices?language=objc)

Hopefully it's just a case of setting up these details in the InfoAdditions section of your app descriptor file. Worth also checking the logs from your application via the Console app so that you can see if there are any messages about a network block..?

thanks

**Reply by ZhorovEgor** [reply]
As I understand it, the Adobe Animate application in conjunction with the AIR SDK cannot give additional permissions inside their interface. Therefore, I manually registered the innovations in the XML file. Let's look at it.
[XML.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/15470766/XML.txt)

In addition, you can throw off an example of an application descriptor for adobe animate so that the permission to use networks is registered there

**Reply by ajwfrost** [reply]
Yes, correct, you would need to add it manually to the XML -> looks about right from what I can see there..

If that still doesn't work then please let us know what AS3 code you're calling here when it fails, and whether there's anything shown up in the console when it blocks it... we'd need to be able to reproduce the problem here I think.

thanks

**Reply by ZhorovEgor** [reply]
The application does not fail, the application simply does not have access to local networks. there is only access to cellular phones, my task is to open access to the application through the application descriptor. (similar to android)

thanks

**Reply by ZhorovEgor** [reply]
Hello 👋  How about my question?

**Reply by marchbold** [reply]
Are you talking about the privacy permissions around local network access in iOS 14+?  Are you looking to use ip multicast or broadcast? 

You will need to add the `NSLocalNetworkUsageDescription` to info additions:

```xml
<key>NSLocalNetworkUsageDescription</key>
<string>This app uses the local network to search for services</string>
```

And maybe some additions for services, depending on what you are using?

```xml
<key>NSBonjourServices</key>
<array>
    <string>_http._tcp</string>
</array>
```

Also I think you need to remove `NSAllowsArbitraryLoads` if you are using this.

**Reply by ZhorovEgor** [reply]
> Are you talking about the privacy permissions around local network access in iOS 14+?  Are you looking to use ip multicast or broadcast? 
> 
> You will need to add the `NSLocalNetworkUsageDescription` to info additions:
> 
> ```xml
> <key>NSLocalNetworkUsageDescription</key>
> <string>This app uses the local network to search for services</string>
> ```
> 
> And maybe some additions for services, depending on what you are using?
> 
> ```xml
> <key>NSBonjourServices</key>
> <array>
>     <string>_http._tcp</string>
> </array>
> ```
> 
> Also I think you need to remove `NSAllowsArbitraryLoads` if you are using this.

My application developed for
smart stairs. I need to connect to his controller. How you know, AS3 app it is cross-platform, my app was compilation on AIR, and publish in Play market https://play.google.com/store/apps/details?id=air.WizDom 

Idk why the same app do not wants to connect to the internet on IOS

**Reply by ZhorovEgor** [reply]
Hello guys,  what I can do?

**Reply by ZhorovEgor** [reply]
> Are you talking about the privacy permissions around local network access in iOS 14+?  Are you looking to use ip multicast or broadcast? 
> 
> You will need to add the `NSLocalNetworkUsageDescription` to info additions:
> 
> ```xml
> <key>NSLocalNetworkUsageDescription</key>
> <string>This app uses the local network to search for services</string>
> ```
> 
> And maybe some additions for services, depending on what you are using?
> 
> ```xml
> <key>NSBonjourServices</key>
> <array>
>     <string>_http._tcp</string>
> </array>
> ```
> 
> Also I think you need to remove `NSAllowsArbitraryLoads` if you are using this.

I just saw app XML, and <key>NSLocalNetworkUsageDescription</key> is used on file.

**Reply by ajwfrost** [reply]
Could you please:
- connect the phone to your mac and run the "Console" app, filtering for your process name, and from this see whether there are any warnings about the network access at the point when you would expect it to connect out?
- confirm what actual AS3 calls you are making that are not working because of this behaviour?

I think we need to try to reproduce it here, to understand what's happening and why it might be blocked..

thanks

**Reply by ZhorovEgor** [reply]
Hello guys, can you get some ready XML file, when local internet is using?

**Reply by ZhorovEgor** [reply]
Good day! Can you tell me how I can upload my info.plist file on my own? If the environment doesn't allow me to create it myself?
![Изображение WhatsApp 2024-06-09 в 10 31 24_bee6d83d](https://github.com/airsdk/Adobe-Runtime-Support/assets/63095086/9f0e3633-4a34-4398-b0d8-fcfc443d1b0c)

**Reply by FliplineStudios** [reply]
You can't add your own Info.plist to your app, AIR will automatically generate it, and it will add items from the `<iPhone><InfoAdditions>` section of your application XML to the Info.plist it generates.  You can double-check what it has added by unzipping your IPA and opening the Info.plist that was created.

**Reply by ZhorovEgor** [reply]
Indeed, I opened the **IPA** file in this way and it seemed to me that the info.plist file was created correctly. But I do not understand why, at startup, the application does not require permission to access the **_local network_**? Considering the previously provided **TXT** file with an example of my XML configuration?

**Reply by FliplineStudios** [reply]
Not really familiar with local networking, but it sounds like iOS may handle things differently based on which version of iOS the device is running:
https://developer.apple.com/documentation/bundleresources/information_property_list/nsapptransportsecurity/nsallowslocalnetworking

>In iOS 17, iPadOS 17, and macOS 14, ATS no longer allows connections to IP addresses by default. Add individual IP addresses and classless inter-domain routing (CIDR) ranges in the NSExceptionDomains dictionary.

I would imagine the permission pop-up for the local network would pop up when you first request a connection, but maybe something is preventing that from happening.  Like @ajwfrost mentioned earlier, I'd think your best bet would be to connect to a Mac and use the "Console" app to see what kind of error messages iOS is outputting when the connection fails, to help with setting this up and debugging.

**Reply by marchbold** [reply]
@FliplineStudios Yeah those are the additions I listed above for the local network access permission dialog and protocols. You need to include them to get local access to work. 

@ZhorovEgor Did you try adding those and the protocols that you are looking to use over the local network?
